### DISCUSSION TITLE
How to resolve own hlasm Macro without host connection

### DISCUSSION THREAD

**Discussion by Eggstoi** [IBM/zopeneditor-about#262] [Q&A]
We are working with z/VSE, so we don't have  ZOWE connected to Mainframe. The source files are manually copied to local directories. That's working fine and also hlasm copybooks are resolved correctly.

But we still have a lot of hlasm macros and also would like to get them resolved. Is it possible to get it working?

Best regards
Andi

**Comment by FALLAI-Denis** [comment]
Hi @Andi82ka 



Regarding the resolution of assembly macros, my tests lead me to suppose that Z Open Editor does not know how to process them, in fact does not identify them in the source code because a macro is written like any assembly instruction, unlike a copybook which has a specific call syntax: `COPY copyname`

I have a configuration with z/OSMF, I declared the access paths to the assembler macros in the zapp.yaml file, however Z Open Editor does not process them.

Concerning the local availability of assembly copybook: is it a need close to #218?
* Some Zowe CLI plugins require a server like z/OSMF, Zowe Server (gateway) or RSE.
* Others are independent of these servers. This is the case of `@zowe/zos-ftp-for-zowe-cli` which only requires an FTP server.
* So if you have an FTP server on z/VSE, then you have a way to retrieve your copies locally thanks to the [VS Code Zowe Explorer Extension for FTP](https://marketplace.visualstudio.com/items?itemName=Zowe.zowe-explorer-ftp-extension).

Next, Z Open Editor would have to support searching for remote information using the FTP protocol, and I would also like it to do so with the SSH protocol, but in either case, as it stands, there is will have character encoding conversion issues:
* Zowe CLI FTP did not plan to manage character conversions while it is provided in the protocol...
* SSH by nature only converts between ASCII ISO-8859-1 and EBCDIC 1047

**Comment by phaumer** [comment]
Yes, we started adding the default macros support as a kind of MVP to see what feedback we would get. We definitely discussed adding custom macros and local macros support, but wanted to wait that someone asked for it.
